codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -130,184 +132,195 @@ class KotlinJmespathExpressionVisitor(
             unwrapExpr?.let { append(ensureNullGuard(member, it)) }
         }
 
-        currentShape = member
-        return addTempVar(name, codegen) to currentShape
+        shapeCursor.addLast(member)
+        return VisitedExpression(addTempVar(name, codegen), member)
     }
 
-    override fun visitAnd(expression: AndExpression): Pair<String, Shape?> {
+    override fun visitAnd(expression: AndExpression): VisitedExpression {
         writer.addImport(RuntimeTypes.Core.Utils.truthiness)
 
-        val (leftName, _) = acceptSubexpression(expression.left)
-        val leftTruthinessName = addTempVar("${leftName}Truthiness", "truthiness($leftName)")
+        val left = acceptSubexpression(expression.left)
+        val leftTruthinessName = addTempVar("${left.identifier}Truthiness", "truthiness(${left.identifier})")
 
-        val (rightName, _) = acceptSubexpression(expression.right)
+        val right = acceptSubexpression(expression.right)
 
-        return addTempVar("and", "if ($leftTruthinessName) $rightName else $leftName") to null
+        val ident = addTempVar("and", "if ($leftTruthinessName) ${right.identifier} else ${left.identifier}")
+        return VisitedExpression(ident, null)
```

Comment:
**Comment**: You return an identifier without a shape often enough that it's probably worth just defaulting the `shape` parameter:

```kotlin
data class VisitedExpression(val identifier: String, val shape: Shape? = null)
```